### PR TITLE
Fix Issue 12507 - SysTime.init.toString should not segfault

### PR DESCRIPTION
**Pull request by n8sh**
Unbelievably, [this bug was reported on April 2, 2014](https://issues.dlang.org/show_bug.cgi?id=12507). I doubt it is of great interest to anyone precisely what is displayed by `SysTime.init.toString`, but it shouldn't cause a segfault.

### CODE REVIEW

File: std/datetime/systime.d [hunk 1]
```diff
@@ -2336,13 +2336,13 @@ public:
...
         timeInfo.tm_yday = dateTime.dayOfYear - 1;
-        timeInfo.tm_isdst = _timezone.dstInEffect(_stdTime);
+        timeInfo.tm_isdst = _timezone ? _timezone.dstInEffect(_stdTime) : 0;
 
         version(Posix)
         {
             import std.utf : toUTFz;
             timeInfo.tm_gmtoff = cast(int) convert!("hnsecs", "seconds")(adjTime - _stdTime);
-            auto zone = (timeInfo.tm_isdst ? _timezone.dstName : _timezone.stdName);
+            auto zone = _timezone is null ? "" : (timeInfo.tm_isdst ? _timezone.dstName : _timezone.stdName);
```
Comment (quickfur) [inline std/datetime/systime.d]: Nitpick: wrap the long line.

**Comment by schveiguy** [comment]
Technically, this should be fixed by making the default timezone printable. But that ran into compiler errors: #5683

**Comment by quickfur** [comment]
Ugh.  I hate compiler bugs. :-(

**Comment by wilzbach** [comment]
CC @jmdavis.
Should we go with this solution as the proper one seems to be out of reach for the foreseeable future?

**Comment by jmdavis** [comment]
> Should we go with this solution as the proper one seems to be out of reach for the foreseeable future?

Since I have no idea how to get anyone to actually fix the bug blocking the correct solution, and nothing I've tried thus far has managed to work around the bug, we probably should go with a worse solution in the short term.

However, these changes don't cover all of the places that `_timezone` is used, and I really think that some comments are required so that it will be easier to remove the unnecessary changes when the blocking bug is finally fixed. It will probably be quicker for me to just do the PR myself than to explain what I think should be done. I probably should have done it a while ago, but I keep hoping that the blocking bug will be fixed, and I hate adding null checks to fix this. Unfortunately, that `Rebindable` in `SysTime` has historically run into a number of bugs.

I'll put something together and either have it up in a little bit here or in the morning.

**Comment by n8sh** [comment]
>It will probably be quicker for me to just do the PR myself than to explain what I think should be done.

Pushing to this PR is enabled for this reason. No need to waste time explaining small things, just change them.

**Comment by jmdavis** [comment]
I went with a hacked up version of the "correct" solution that introduces `InitTimeZone` for `SysTime.init` but uses a helper function that does the null check everywhere that a function would be called on `_timezone` and returns `InitTimeZone()` instead. Once the compiler bug has been fixed, the helper function can be removed, but in the interim, that's the closest that I can get to implementing the "correct" solution given the compiler bug that's blocking it:

https://github.com/dlang/phobos/pull/6591

**Pull request closed without merging**